documentation/Persistent-Problems.md

```diff
@@ -0,0 +1,19 @@
+# Builds suck
+
+No build is ever fast enough, reliable enough, or does everything you want.
+
+MSBuild-driven builds are no exception.
+
+## Evaluation
+
+*Every* [batch build]() (absent a higher-order build system) must evaluate every project in the scope of the build. IDEs can cache evaluation and act as a higher-order builds sytem but important IDE scenarios like “first load after a repo clone” are dominated by evaluation.
+
+## ResolveAssemblyReferences
+
+When build is invoked, most targets can be skipped as up to date, but `ResolveAssemblyReferences` (RAR) and some of its prerequisites like `ResolvePackageAssets` cannot, because their role is to produced data used within the build to compute the compiler command line. Since they don't have concrete file outputs and their file inputs can be difficult to express (it’s the closure of all referenced assemblies), MSBuild's standard up-to-date check mechanisms can't apply.
+
+## Copy
+
+The amount of time spent copying files in a build can be surprising. Efficient copy-on-write filesystems can help dramatically (we now have this on all major operating systems via `clonefile` on [Linux]() and [macOS]() and the [Windows 11 24H2+ Dev Drive]()).
+
+As an implementation detail of MSBuild's common copies, the targets are generally not incremental, in favor of fine-grained incrementality within the Copy task itself. This means that Copy task time can be nonzero even on a fully up-to-date build.
```

Comment:
This 2 sentences are bit loaded - I needed to reread this 3 times to fully understand.
I feel it might be helpful to add a sample of some common target copy target (e.g. `_CopyOutOfDateSourceItemsToOutputDirectory`) and explain parts of it ("No inputs and outputs to the target itself; the target depends on targets that find out what should be copied and what not; and the Copy tasks checks what is unchanged")

Reply:
This behavior is at least in part because we don't have some other process/component that tracks inputs/outputs and have to recreate the Copy sources/destinations every build, right? We on-demand compute the up-to-date check and that can be expensive. Theoretically at least with MSBuild Server we could hang that bookeeping on the server node and _potentially_ track Copy inputs/outputs more cleanly?